### PR TITLE
[stdune] add scanf module

### PR DESCRIPTION
**Pull request by rgrinberg**
This is a slightly safer version that only catches 2 exceptions that
scanf throws

### CODE REVIEW

File: src/dune/build.ml [hunk 1]
```diff
@@ -118,7 +118,13 @@ let contents p = Contents p
 
 let lines_of p = Lines_of p
 
-let strings p = Map ((fun l -> List.map l ~f:Scanf.unescaped), lines_of p)
+let strings p =
+  let f x =
+    match Scanf.unescaped x with
+    | Error () -> Code_error.raise "Build.strings" []
```
Comment (diml) [inline src/dune/build.ml]: This is a file produced by the external world, so this shouldn't be a code error. What about raising a user error located in `p`?

File: src/stdune/scanf.ml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module Scanf = Dune_caml.Scanf
+
+let unescaped x =
+  match Scanf.unescaped x with
+  | exception End_of_file -> Error ()
```
Comment (aalekseyev) [inline src/stdune/scanf.ml]: Why only these two? The doc seems to say:

> Raise Scanf.Scan_failure if the input does not match the format.
> Raise Failure if a conversion to a number is not possible.
> Raise End_of_file if the end of input is encountered while some more characters are needed to read the current conversion specification.
> Raise Invalid_argument if the format string is invalid.

(but I don't know when the other two are raised)

**Comment by diml** [comment]
> That would indeed be preferable. I just tried to preserve existing behaviour as much as possible. I agree we should be a lot more careful with all these exceptions.

The existing behaviour seems wrong though. Since we are touching this code, we might as well clean it up!

File: src/stdune/scanf.ml (hunk 2 above)
Comment (rgrinberg) [reply src/stdune/scanf.ml]: I'll catch failure as well. However, I think that we shouldn't catch `Invalid_argument _`. A bad formatting string is just a bug.

Comment (rgrinberg) [reply src/stdune/scanf.ml]: Of course, this does make the function a bit more dangerous to use as `f` could easily raise `Failure` for other reasons.

Comment (rgrinberg) [reply src/stdune/scanf.ml]: ping @aalekseyev

Comment (diml) [reply src/stdune/scanf.ml]: @aalekseyev is away this week, I'll finish the review for him.

> Of course, this does make the function a bit more dangerous to use as f could easily raise Failure for other reasons.

What about `Scanf.ksscanf`? It seems to be exactly for this case

Comment (diml) [reply src/stdune/scanf.ml]: In fact, if we use `ksscanf` I'd say we don't even need to inspect the exception

Comment (rgrinberg) [reply src/stdune/scanf.ml]: I pushed a commit that uses `ksscanf`. However, I'm unable to accomplish this without this private exception I've defined. Perhaps you can think of a way to do it without this?

File: src/stdune/scanf.ml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+module Scanf = Dune_caml.Scanf
+
+let unescaped x =
+  match Scanf.unescaped x with
+  | exception Scanf.Scan_failure _ -> Error ()
+  | x -> Ok x
+
+let sscanf x fmt f =
+  let module M = struct
+    exception E
```
Comment (diml) [inline src/stdune/scanf.ml]: We should put this at the toplevel given that defining an exception is not cost-free. Indeed it requires generating a global identifier.

File: src/stdune/scanf.ml (hunk 2 above)
Comment (diml) [reply src/stdune/scanf.ml]: The method with a private exception looks good to me

**Comment by rgrinberg** [comment]
@diml now it's a toplevel exception

**Pull request merged**